### PR TITLE
Use new_x and new_y in place of nl=0, retire center=False

### LINKED ISSUE #343: Simplify specifying the new position after multline text
When having a closer look at how to best adapt `.write()` to the new line wrapping code (per #340), I ended up struggling a bit with the concept behind the `ln=#` parameter to both `.multicell()` and `.write()`, and now also `._render_styled_cell_text()`.  Judging by its name, I suspect that originally it was only meant as a `True/False` value to indicate wether the position should move to a new line. Later it seems to have been increasingly overloaded to also indicate horizontal movement (or maybe it was badly designed right from the start, who knows...) Either way, using an arbitrary single numerical value to indicate movement choices in both x and y is rather confusing and unintuitive. 

To improve the clarity of the API and also simplify maintenance, I think it would make sense to seperate the two functionalities. The simplest way is probably to introduce a seperate argument `new_xpos=""`, with a single character value choice of `"L|R|S|E"`, for **L**eft/**R**ight of the requested cell extents, or **S**tart/**E**nd of the actual text rendered. 

The `ln=` parameter could then again be reduced to a `True|False` choice of moving to a new line or not, ideally with a deprecation warning for other uses.

Any thoughts? Better names for the new parameter?

### PR DESCRIPTION
**Pull request by gmischler**
e.g. Fixes #343

**Checklist**:

- [x] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.
- [x] A unit test is covering the code added / modified by this PR
- [x] This PR is ready to be merged 
- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder
- [x] A mention of the change is present in `CHANGELOG.md`
- [x] The PR description or comment contains a picture of a cute animal (not mandatory but encouraged 🐘 )

This one was mostly boilerplate code changes, so it didn't take a lot of time. Hunting down all the uses elsewhere in the code (including the "hidden" ones in html.py) took the most effort.

There was very little mention of ln=# in the tutorials, so I could easily fix those, even in the languages that I can't even read...
While that simplified the task at hand here, maybe an extra documentation page explaining all the text layout options might eventually be helpful.

Edit:
I had to move the definitions of XPos and YPos to util.py, to avoid a circular import issue with html.py.

### CODE REVIEW

File: CHANGELOG.md
Comment (Lucas-C) [inline CHANGELOG.md:12]: This should be added to the `2.5.2` section above.
You should probably rebase your branch on this repo `master` branch

Comment (gmischler) [reply CHANGELOG.md:12]: I thought I had the most recent data... now I actually do!

File: fpdf/fpdf.py
Comment (Lucas-C) [inline fpdf/fpdf.py:1943]: `ln='DEPRECATED'` would be more explicit and more consistent with how other deprecated arguments are currently handled

File: fpdf/fpdf.py
Comment (Lucas-C) [inline fpdf/fpdf.py:2117]: 👍

File: fpdf/html.py
Comment (Lucas-C) [inline fpdf/html.py:309]: Do you think you could use this opportunity to the first element of the 2-tuples in `self.theader` list to be `dict`? This way, in `output_table_footer()`, `self.pdf.cell(**cell)` could be called, and the code would not rely on the ordering of arguments passed to this method, which is a bit brittle.

File: test/text/test_cell.py
Comment (Lucas-C) [inline test/text/test_cell.py:85]: 👍

File: test/text/test_render_styled.py
Comment (Lucas-C) [inline test/text/test_render_styled.py:12]: Could you uppercase this constant?

File: test/text/test_render_styled.py
Comment (Lucas-C) [inline test/text/test_render_styled.py:178]: Could you find more explicit names for those constants? 😅

File: fpdf/fpdf.py
Comment (Lucas-C) [inline fpdf/fpdf.py:2515]: Given how `ln=X` is a lot more shorter than `new_x=XPos.YYY, new_y=YPos.ZZZ`, I suspect that some `fpdf2` users may find the parameters less handy than the original one.
Hence I wonder if we shouldn't allow users to "ignore" this deprecation / replacement.
I'm thinking about 2 strategies:
* ensure this warning is only raised **once**, so that users can safely ignore it without having their terminal / logs being polluted by tens of warning lines
* subtype `DeprecationWarning` in this case, and suggest to users who would like to continue using `ln` to call ` warnings.simplefilter('ignore', LnArgDeprecationWarning)`.

What do you think about those options @gmischler ? Should we adopt both?

Comment (gmischler) [reply fpdf/fpdf.py:2515]: According to the documentation, the default warnings filter only prints each message once per location it was raised from, so the first point is already a given.

Seeing that the new paramaters are much more intuitive, memorable, and precise, I somehow doubt that people will refuse to use them just to save a few keystrokes once. If you want users to be able to treat different warnings differently, then it might make more sense to establish a more systematic hierarchy, than to single out this specific case.

But generally speaking, don't we want them to actually abandon the deprecated features as soon as possible? Unfortunately, that requires to get on their nerves at least a little bit. If you make it too easy to just ignore specific warnings, then they'll get bitten by them much harder once you actually remove the cruft. Or how long do you plan to keep obsolete APIs around?

File: fpdf/html.py
Comment (gmischler) [reply fpdf/html.py:309]: I tried to touch this part as little as possible for the moment... But that is indeed a very good idea in this case (especially when using 'DEPRECATED' instead of `None`).

File: test/text/test_render_styled.py
Comment (gmischler) [reply test/text/test_render_styled.py:12]: done, with a more speaking name even.

File: test/text/test_render_styled.py
Comment (gmischler) [reply test/text/test_render_styled.py:178]: already done due to the notice above... :wink:

File: fpdf/fpdf.py
Comment (Lucas-C) [reply fpdf/fpdf.py:2515]: > According to the documentation, the default warnings filter only prints each message once per location it was raised from, so the first point is already a given.

Nice! I didn't realize that.

> But generally speaking, don't we want them to actually abandon the deprecated features as soon as possible? Unfortunately, that requires to annoy them at least a little bit.

There is a good read about deprecation policies: https://steve-yegge.medium.com/dear-google-cloud-your-deprecation-policy-is-killing-you-ee7525dc05dc

> (quote) Some folks believe that [library] developers should shoulder the costs of compatibility, and others believe that platf[library] users should bear the costs.

What I'd like for `fpdf2` is to provide a clear API for newcomers, but at the same time avoid creating trouble for existing users. I really think we should find a way to introduce clean, clear parameters while limiting any "annoyance" as much as possible 😊

> If you make it too easy to just ignore specific warnings, then they'll get bitten by them much harder once you actually remove the cruft. Or how long do you plan to keep obsolete APIs around?

For the reasons explained in that article, probably for a long time!

But I agree that we don't want to "fall between two stools" here. I think that if I'm reluctant to force users to switch to your new `new_x=XPos.YYY, new_y=YPos.ZZZ` syntax that's simply because I am not fully convinced by it yet...
It is quite verbose to write & read compared to the original `ln=X` system, even if it was very obscure.
But I also see its benefits.

I'm going to leave ourselves some time to think before merging this.

And also, we will need to add some clear documentation about this anyway.

Comment (gmischler) [reply fpdf/fpdf.py:2515]: That article talks about "platforms" (cloud services, Emacs, Android, Java, web browsers). And in that realm, I mostly agree with the author: Making stuff disappear can be a problem there, primarily because the user has little control over which version they can use and what third-party components it needs to work with.

Fpdf is almost the exact opposite of such a platform, so the same conclusions don't necessarily apply here.
It is a "minimalistic" library, that is usually combined as a small part within a larger software system.
Developers are free to decide if and when they are going to upgrade. And by the very nature of the upgrade concept, they will expect changes when they do that. Anyone who doesn't like a new version for whatever reason, will be able to keep running the old one pretty much forever.

If a developer publishes their stuff eg. via pip, they can (and should!) declare the **tested to be compatible** library versions in the dependency manifest. If a library feature disappears in a future version, the developers who followed that rule won't even notice until they voluntarily upgrade, and their end users will never notice at all.

Due to its history, fpdf has accumulated a lot of technical debt, as a concequence of some suboptimal decisions that were made in the distant past. Parts of its API have been cobbled together rather haphazardly and will eventually need to be fixed.

This "ln" parameter is a prime example for that. It's quite obvious that at first it was simply a 1|0 switch: "should cell() jump to the next line or not." Then somone decided to overload it with additional semantics and also use it to control the horizontal offset. And on top of that, when the concept was transferred to multi_cell(), the meaning of the value "0" was changed to mean "next line" instead of "same line". 

The result makes no sense at all. Even after working on it for a while, trying to replace it as seamlessly as possible, I still can't memorize which value means what. Any sane user should be happy to see that nonsense disappear.

If you want fpdf to remain "minimalistic", then you must get rid of obsolete cruft as soon as humanly possible. This means you need a deprecation strategy **with a timeline**. At the moment, "deprecated" just means "may go away eventually". This seems way too vague, and there's no good reason to be vague here.

"Explicit is better than implicit."
My reccommendation would be, whenever you declare something as deprecated, you also immediately decide for how long it will be staying around, and include that information in the warning message:
```
DeprecationWarning: Use "something" instead. Will be removed in version X.Y.
```
And the closer you get to X.Y, the more annoying you can make that warning. 
Declaring a timeline explicitly like that best satisfies the principle of "least surprise".

I would set the target version to something that you expect to be between 6 and 12 months away (based on the recent update cadence). But you can vary that depending on the severity of the change or your personal taste.

File: CHANGELOG.md
Comment (Lucas-C) [inline CHANGELOG.md:17]: This is redundant with the next line

File: docs/Text.md
Comment (Lucas-C) [inline docs/Text.md:1]: This PR contains commits that are part of https://github.com/PyFPDF/fpdf2/pull/363 ?

File: fpdf/fpdf.py
Comment (Lucas-C) [inline fpdf/fpdf.py:1946]: Could you also change the default value of the `center` parameter to `"DEPRECATED"`?
And raise a warning in this method, so that we don't pass it anymore to `_render_styled_cell_text()`

File: fpdf/fpdf.py
Comment (Lucas-C) [inline fpdf/fpdf.py:2622]: The intent here is to uncomment that section in a later PR?

**Review by Lucas-C — APPROVED** [review]
@gmischler : I'm ok to merge this once we have dealt with the final minor changes I suggested.

Also, I got an idea to limit the verbosity of the `new_x=XPos.YYY, new_y=YPos.ZZZ` syntax, that we could potentially implement in a later PR.
What would you think about introducing a new method `FPDF.set_flow_mode(new_x=XPos.YYY, new_y=YPos.ZZZ)` that would set the default value of `new_x` & `new_y` for all methods that accept those parameters?

File: docs/Text.md
Comment (gmischler) [reply docs/Text.md:1]: I started that here and later decided to make it a separate PR.
Will remove the leftovers here.

File: fpdf/fpdf.py
Comment (gmischler) [reply fpdf/fpdf.py:1946]: Ah yes, of course.
The warning gets raised, but we still pass any values on. It's not formally removed yet, after all.

File: fpdf/fpdf.py
Comment (gmischler) [reply fpdf/fpdf.py:2622]: Probably. I'll remove it for now.

**Comment by gmischler** [comment]
> Also, I got an idea to limit the verbosity of the `new_x=XPos.YYY, new_y=YPos.ZZZ` syntax, that we could potentially implement in a later PR. What would you think about introducing a new method `FPDF.set_flow_mode(new_x=XPos.YYY, new_y=YPos.ZZZ)` that would set the default value of `new_x` & `new_y` for all methods that accept those parameters?

So you want to add yet more global state to the library, as well as another metod call, all in the name of simplicity?
You can't really be serious about that, can you?
Note that those options are quite likely to change from call to call, and the defaults between `cell()` and `multi_cell()` are diffent, while `write()` handles them internally. Turning something like that into a global state could only make sense if the goal was to cause maximal confusion.
To be honest, I think your worries about verbosity are a bit exaggerated here. I see no good way to make the necessary information less verbose other than shortening the names.

**Comment by gmischler** [comment]
So after returning home, I finally managed to apply the last fixes.
@Lucas-C, Holler if I missed something.

**Review by Lucas-C — APPROVED** [review]

**Comment by Lucas-C** [comment]
Merged! Thank you @gmischler 😊

**Pull request merged**
